exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Queries.kt

```diff
@@ -492,18 +501,21 @@ fun <T : Table, E : Any> T.batchUpsert(
  * primary keys will be used. If the table does not have any primary keys, the first unique index will be attempted.
  * @param onUpdate List of pairs of specific columns to update and the expressions to update them with.
  * If left null, all columns will be updated with the values provided for the insert.
+ * @param onUpdateExclude Set of specific columns to exclude from updating.
+ * If left null, all columns will be updated with the values provided for the insert.
  * @param shouldReturnGeneratedValues Specifies whether newly generated values (for example, auto-incremented IDs) should be returned.
  * See [Batch Insert](https://github.com/JetBrains/Exposed/wiki/DSL#batch-insert) for more details.
  * @sample org.jetbrains.exposed.sql.tests.shared.dml.UpsertTests.testBatchUpsertWithNoConflict
  */
 private fun <T : Table, E> T.batchUpsert(
     data: Iterator<E>,
-    vararg keys: Column<*>,
     onUpdate: List<Pair<Column<*>, Expression<*>>>? = null,
+    onUpdateExclude: Set<Column<*>>? = null,
     shouldReturnGeneratedValues: Boolean = true,
-    body: BatchUpsertStatement.(E) -> Unit
+    body: BatchUpsertStatement.(E) -> Unit,
+    vararg keys: Column<*>
```

Comment:
Parameter order is changed here to put `vararg` last, so that calls to this private function don't need every single argument to be named (see diff in both `batchUpsert()` above).

Reply:
Could we still keep `body` as a last parameter so it can be used as dsl syntax?